akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/AccountExample3b.scala

```diff
@@ -0,0 +1,117 @@
+        case _              ⇒ throw new IllegalStateException(s"unexpected event [$event] in state [EmptyAccount]")
+      }
+    }
+    case class OpenedAccount(balance: Double) extends Account {
+      override def applyEvent(event: AccountEvent): Account = event match {
+        case Deposited(amount) ⇒ copy(balance = balance + amount)
+        case Withdrawn(amount) ⇒ copy(balance = balance - amount)
+        case AccountClosed     ⇒ ClosedAccount
+        case _                 ⇒ throw new IllegalStateException(s"unexpected event [$event] in state [OpenedAccount]")
+      }
+    }
+    case object ClosedAccount extends Account {
+      override def applyEvent(event: AccountEvent): Account =
+        throw new IllegalStateException(s"unexpected event [$event] in state [ClosedAccount]")
+    }
+
+    def behavior(accountNumber: String): Behavior[AccountEntity.AccountCommand] =
+      Behaviors.setup(ctx ⇒ new AccountEntity(ctx, accountNumber))
+  }
+
+  class AccountEntity(context: ActorContext[AccountEntity.AccountCommand], accountNumber: String)
+    extends PersistentBehavior3[AccountEntity.AccountCommand, AccountEntity.AccountEvent, AccountEntity.Account](
+      persistenceId = accountNumber) {
+    import AccountEntity._
+
+    override protected def emptyState: Account = EmptyAccount
+
+    private def initialHandler(cmd: AccountCommand): Effect[AccountEvent, Account] =
+      cmd match {
+        case CreateAccount ⇒ Effect.persist(AccountCreated)
+        case _             ⇒ Effect.unhandled
+      }
+
+    private def openedAccountHandler(acc: OpenedAccount, cmd: AccountCommand): Effect[AccountEvent, Account] = {
+      cmd match {
+        case Deposit(amount) ⇒ Effect.persist(Deposited(amount))
+
+        case Withdraw(amount) ⇒
+          if ((acc.balance - amount) < 0.0)
+            Effect.unhandled // TODO replies are missing in this example
+          else {
+            Effect
+              .persist(Withdrawn(amount))
+              .thenRun {
+                case OpenedAccount(balance) ⇒
+                  // do some side-effect using balance
+                  println(balance)
+                case _ ⇒ throw new IllegalStateException
+              }
+          }
+        case CloseAccount if acc.balance == 0.0 ⇒
+          Effect.persist(AccountClosed)
+
+        case CloseAccount ⇒
+          Effect.unhandled
+      }
+    }
+
+    private def closedHandler(cmd: AccountCommand): Effect[AccountEvent, Account] = {
+      Effect.unhandled
+    }
+
+    override protected def onCommand(state: Account, cmd: AccountCommand): Effect[AccountEvent, Account] = {
+      state match {
+        case EmptyAccount              ⇒ initialHandler(cmd)
+        case opened @ OpenedAccount(_) ⇒ openedAccountHandler(opened, cmd)
+        case ClosedAccount             ⇒ closedHandler(cmd)
+      }
+    }
+
+    override protected def onEvent(state: Account, event: AccountEvent): Account =
+      state.applyEvent(event)
+
+  }
+
+}
```

Comment:
I think one problematic aspect of this is that it if not encourages, at least leaves a big opening for thinking about the instance as the-behavior and putting additional state outside of the eventsourced state, as regular fields in the class - because that's how the untyped Actors and the `MutableBehavior` work.

Reply:
exactly, all state of importance must be in the `State`

Reply:
That is an excellent point, @johanandren.
I also like this approach for the Scala API. If one starts with the behavior (`PersistentBehaviors.receive`) then the API is quite guided. But should be clear in the docs what the proposed workflow is.

Reply:
@jboner could you clarify what you refer to with "this approach"? This comment is on AccountExample3b, which is the class based approach. I interpret `PersistentBehaviors.receive` as you prefer AccountExample5 or later.

Reply:
Sorry. Too many comments. Yes, I meant AccountExample5b.